File: src/domain/outcome.zig
//! Per-Mutation Apply outcome — the typed result returned by
//! `Adapter.applyMutation` and consumed by the sync engine when it persists
//! `mutations.state` and `mutations.failure_json`.
//!
//! Lives in `src/domain/` because it is pure data with no SQLite, filesystem,
//! Git, or subprocess dependencies. `Receipt` is empty in ticket-17; the
//! Promote slice grows it with backend-assigned identifiers returned by a
//! successful `promote_*` Mutation.

const std = @import("std");
const Allocator = std.mem.Allocator;

/// Distinguishes per-Mutation success from per-Mutation rejection.
///
/// Environment failures (`ExecutableNotFound`, `SpawnFailed`, `OutOfMemory`)
/// are surfaced through the bare `ApplyError` union, not through this tagged
/// union — that split is documented in `docs/adr/0009-sync-failure-taxonomy.md`.
pub const Outcome = union(enum) {
    /// Mutation accepted by the backend.
    success: Receipt,
    /// Mutation rejected by the backend (non-zero exit, validation refusal).
    /// See `Failure.detail` for the ownership contract on the captured slice.
    failure: Failure,
};

/// Adapter-supplied evidence that a Mutation succeeded.
///
/// Intentionally empty for ticket-17 — the Promote slice grows it with the
/// backend-assigned identifiers (e.g. issue number, Jira key) returned by a
/// successful `promote_*` Mutation.
pub const Receipt = struct {};

/// Adapter-supplied evidence that a Mutation was rejected.
pub const Failure = struct {
    /// Human-readable failure detail captured from the adapter (typically the
    /// backend CLI's stderr).
    ///
    /// Ownership contract: the adapter MUST allocate `detail` via the per-call
    /// `gpa` argument passed to `applyMutation`. If copying from a runner-
    /// owned slice such as `proc.Result.stderr`, the adapter must
    /// `gpa.dupe` before returning — the runner's `Result.deinit(gpa)` frees
    /// its own stderr buffer before the engine writes `detail` to
    /// `mutations.failure_json`. The engine takes ownership on return and
    /// frees via a `defer` block.
    detail: []const u8,

    /// Free `detail` with the allocator passed to `applyMutation`.
    pub fn deinit(self: Failure, gpa: Allocator) void {
        gpa.free(self.detail);
    }
};

test "Outcome.failure.deinit frees detail" {
    const gpa = std.testing.allocator;
    const detail = try gpa.dupe(u8, "backend rejected mutation");
    const failure = Failure{ .detail = detail };
    failure.deinit(gpa);
}

File: crates/tk/src/domain/outcome.rs
//! Per-Mutation Apply outcome — the typed result a Backend Adapter returns
//! from `apply_mutation` and the sync engine consumes when it persists
//! `mutations.state` and `mutations.failure_json`.
//!
//! Pure data: no SQLite, filesystem, Git, or subprocess dependencies.
//!
//! This is one of the two typed shapes ADR-0018 deferred until "real adapter
//! pressure" — that pressure is the sync engine (ADR-0003 Mutation outbox
//! replay), which needs to distinguish three levels of result:
//!
//! - environment failure (adapter unavailable: binary missing, spawn failed)
//!   — surfaces through the adapter method's `Result<_, ApplyError>` error
//!   arm, bubbles out of the engine, leaves the Mutation row `pending`;
//! - per-Mutation rejection (the backend ran but refused the write) —
//!   surfaces as [`Outcome::Failure`], recorded to `failure_json`, stops the
//!   apply loop;
//! - acceptance — surfaces as [`Outcome::Success`], transitions the row to
//!   `applied` and advances the Sync Cursor.
//!
//! The success/rejection split (this type) vs. environment failure (the error
//! arm) is the ADR-0009 sync failure taxonomy.

/// Result of handing one Mutation Log entry to a Backend Adapter.
///
/// Environment failures are NOT modelled here — they arrive through the
/// adapter method's `Result` error arm (`ApplyError`). This enum only
/// distinguishes backend acceptance from backend rejection.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Outcome {
    /// Mutation accepted by the backend.
    Success(Receipt),
    /// Mutation rejected by the backend (non-zero exit, validation refusal).
    Failure(Failure),
}

/// Adapter-supplied evidence that a Mutation succeeded.
///
/// Intentionally empty today — the Promote slice grows it with the
/// backend-assigned identifiers (issue number, Jira key) a successful
/// `promote_*` Mutation returns. Kept as a struct rather than a unit variant
/// so that growth is an additive field change, not a variant-shape churn.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default)]
pub struct Receipt {}

/// Adapter-supplied evidence that a Mutation was rejected.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Failure {
    /// Human-readable failure detail captured from the adapter (typically the
    /// backend CLI's stderr). The engine persists this verbatim into the
    /// `{"detail":"…"}` wrapper stored in `mutations.failure_json`.
    pub detail: String,
}

impl Outcome {
    /// Convenience constructor for the empty-receipt success case.
    #[must_use]
    pub fn success() -> Self {
        Self::Success(Receipt::default())
    }

    /// Convenience constructor for a rejection carrying `detail`.
    #[must_use]
    pub fn failure(detail: impl Into<String>) -> Self {
        Self::Failure(Failure {
            detail: detail.into(),
        })
    }
}
